networkx/algorithms/tsp.py

```diff
@@ -0,0 +1,319 @@
+    >>> sol = nx.greedy_tsp(G, 'D')
+    >>> sol[0]
+    ['D', 'C', 'B', 'A', 'D']
+    >>> sol[1]
+    31.0
+
+    Notes
+    -----
+    Implementation of greedy algorithm is based on the following:
+    - Algorithm adds a node to the solution at every
+    iteration.
+    - Algorithm selects a node whose connection with the previous node
+    adds the minimum cost to incomplete solution at every iteration.
+
+    A greedy algorithm does not give always the best solution.
+    However, it can construct a first feasible solution which can
+    be passed as parameter in iterative improvement algorithm such
+    as Simulated Annealing, Threshold Accepting.
+    """
+    if not _is_completed(G):
+        raise nx.NetworkXError('Given graph is not completed.')
+    if not nx.is_weighted(G, weight=weight):
+        raise nx.NetworkXError('Given graph is not weighted.')
+    nodelist = G.nodes()
+    nodelist.remove(source)
+    sol = [source]
+    cost = 0.0
+    while len(nodelist) > 0:
+        next_visitor, dist = _select_next(G, source, nodelist, weight)
+        sol.append(next_visitor)
+        cost += dist
+        nodelist.remove(next_visitor)
+        source = next_visitor
+    sol.append(sol[0])
+    cost += G.edge[sol[-2]][sol[0]][weight]
+    return sol, cost
+
+
+def _is_completed(G):
+    """
+    Function to test if a graph is completed.
+
+    :param G: A networkX Graph
+    :return: True if graph is completed; False otherwise.
+    """
+    return all(G.has_edge(u, v) for u in G.nodes() for v in G.nodes() if u != v)
+
+
+def _select_next(G, source, nodes, weight='weight'):
+    """
+    Function which selects the next node where salesman has to
+    visit.
+
+    Node whose connection with the source node given as parameter
+    has the minimum weight amongst others.
+
+    :param G: G: A networkX Graph.
+    :param source: Node, source node.
+    :param nodes: List of nodes which have not been visited yet.
+    :param weight: Edge data key corresponding to the edge weight.
+    :return: Tuple, edge of source node and next node to visit.
+    """
+    visitors = [(node, G.edge[source][node][weight])
+                for node in nodes]
+    return sorted(visitors, key=itemgetter(1))[0]
+
+
+def simulated_annealing_tsp(G, source, temp=100, move='1-1', outer_iter=10,
+                            inner_iter=100, a=0.01, sol=None, weight='weight'):
+    """Finds the route that salesman has to visit in order
+    to minimize total distance and total distance using a
+    simulated annealing algorithm.
+
+    In essence, function returns a cycle given a source point
+    which total cost of cycle is minimized.
+
+    Parameters
+    ----------
+    G: A complete weighted graph.
+
+    source : node
+        Starting node
+
+    temp : int, optional (default=100)
+        An algorithm's parameter. It represents the initial
+        temperature of algorithm
+
+    move : string, optional (default='1-1')
```

Comment:
What are the possible values for `move`? What do these represent? Needs to be documented better.